File: android/src/main/java/com/navigine/naviginesdk/CircleMapObjectController.java
package com.navigine.naviginesdk;

import android.content.Context;

import com.navigine.idl.java.CircleMapObject;
import com.navigine.idl.java.LocationWindow;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;

public class CircleMapObjectController {
  private final BinaryMessenger binaryMessenger;
  private final Context context;
  private final LocationWindow locationWindow;
  private final Map<Integer, NavigineCircleMapObject> mapObjects = new HashMap<>();

  @SuppressWarnings({"ConstantConditions", "unchecked"})
  public CircleMapObjectController(
          BinaryMessenger messenger,
          Context context,
          LocationWindow locationWindow
  ) {
    this.binaryMessenger = messenger;
    this.context = context;
    this.locationWindow = locationWindow;
  }

  public int addCircleMapObject() {
    CircleMapObject circleMapObject = locationWindow.addCircleMapObject();
    int circleMapObjectId = circleMapObject.getId();
    NavigineCircleMapObject navigineCircleMapObject = new NavigineCircleMapObject(
            circleMapObjectId,
            this.context,
            circleMapObject,
            binaryMessenger);

    mapObjects.put(circleMapObjectId, navigineCircleMapObject);
    return circleMapObjectId;
  }

  public boolean removeCircleMapObject(int id) {
    NavigineCircleMapObject navigineCircleMapObject = mapObjects.get(id);
    if (navigineCircleMapObject == null) {
      return false;
    }

    boolean success = locationWindow.removeCircleMapObject(navigineCircleMapObject.getMapObject());
    mapObjects.remove(id);
    return success;
  }


}

File: ios/Classes/CircleMapObjectController.swift
import Navigine

class CircleMapObjectController: NSObject {
    public weak var locationView: NCLocationView?

    private let pluginRegistrar: FlutterPluginRegistrar!
    private var mapObjects: [Int32: NavigineCircleMapObject] = [:]

    public required init(
        registrar: FlutterPluginRegistrar,
        locationView: NCLocationView
    ) {
        self.pluginRegistrar = registrar
        self.locationView = locationView
        super.init()
    }

    public func addCircleMapObject() -> Int32? {
        if locationView == nil {
            return nil
        }
        let circleMapObject = locationView!.addCircleMapObject()
        let circleMapObjectId = circleMapObject.getId()
        mapObjects[circleMapObjectId] = NavigineCircleMapObject(id: circleMapObjectId, mapObject: circleMapObject, registrar: self.pluginRegistrar)
        return circleMapObjectId

    }

    public func removeCircleMapObject(id: Int32) -> Bool? {
        if locationView == nil {
            return nil
        }
        let circleMapObject = mapObjects[id]
        if (circleMapObject == nil) {
            return false
        }

        let success = locationView?.remove((circleMapObject?.getMapObject())!);
        mapObjects.removeValue(forKey: id)
        return success
    }
}
